Adding language tests

arnavkapoor opened the pull request:
So I have added / reformatted the tests for ``` en, es and ru```. Also the parser has some updates to correctly handle multipliers, so the logic has become a bit more complex to handle all the cases. There are definitely some ambiguous cases , and I have added tests for the same in ```test_ambiguity_in_multipliers``` in the ```test_language_en.py``` file would appreciate all opinions on the expected output for these cases. 

It's still a draft as I was thinking of incorporating the following in this PR
- Hindi language tests.
- Adding the supplementary data to handle the failing cases in es and ru. ( As discussed would treat millón/millones and cien/ciento the same for the time-being)

noviluni commented on tests/test_language_ru.py:
```diff
@@ -0,0 +1,80 @@
+            (21000000, "двадцать одним миллионом"),
+        ]
+    )
+    def test_parse_number_advanced(self, expected, test_input):
```
I prefer to have a "common interface" to add the tests, with the same test naming in each language test file. As this test is checking the same function as `test_parse_number_basic`, what do you think about putting all of them together by adding a comment in the middle (and rename `test_parse_number_basic` --> `test_parse_number`) ?

Example:
```python
...
(7004, "семь тысяч четыре"),
(21000000, "двадцать один миллион"),
(21000073, "двадцать один миллион семьдесят три"),
# Advanced cases
(1, "одного"),
(1, "одному"),
(1, "одним"),
...
```

noviluni commented on tests/test_language_en.py:
```diff
@@ -0,0 +1,145 @@
+                "the     sun      is     112 km     away."),
+        ]
+    )
+    def test_basic_sentences(self, expected, test_input):
```
It's a good pattern to write the name of the tested function after the `test_`. I would rename this function to:

```suggestion
    def test_parse_basic_sentences(self, expected, test_input):
```

The same for the other tests: 
* `test_case_of_string` --> `test_parse_case_of_string`
* `test_ambiguity_in_multipliers` --> `test_parse_ambiguity_in_multipliers`
* `test_ambiguity_in_separators` --> `test_parse_ambiguity_in_separators`

and so on.

noviluni commented on tests/test_language_es.py:
```diff
@@ -0,0 +1,56 @@
+        ]
+    )
+    def test_parse_number(self, expected, test_input):
+        assert parser.parse_number(test_input, LANG) == expected
```
Apart from adding a `test_parse_numer` test, I would add a `test_parse` test with at least three sentences for each language.

Here some examples for Spanish. Let me know if they don't work:

* "Hay tres gallinas y veintitrés patos" --> "Hay 3 gallinas y 23 patos"
* "En España viven cuarenta y seis millones novecientas cuarenta mil personas". --> "En España viven 46940000 personas".
* "dos y dos son cuatro cuatro y dos son seis seis y dos son ocho y ocho dieciséis" --> "2 y 2 son 4 4 y 2 son 6 6 y 2 son 8 y 8 16".  # this is specially interesting as it has some numbers with "y" that shouldn't be joined.

noviluni commented on tests/test_number_parsing.py:
```diff
@@ -1,107 +0,0 @@
-from number_parser import parser
-
-class TestNumberParser():
```
I will leave this file to add tests that don't rely on a specific language.

noviluni commented on tests/test_language_ru.py:
```diff
@@ -0,0 +1,80 @@
+            (256, "двести пятьдесят шесть"),
+            (256, "двухсот пятидесяти шести"),
+            (256, "двумястами пятьюдесятью шестью"),
+            (21000000, "двадцать одним миллионом"),
```
I think some tests here are failing and should be marked with `xfail`, right? If it's too hard to write the reason don't write it :+1:

noviluni commented on number_parser/parser.py:
```diff
@@ -109,16 +131,16 @@ def parse_number(input_string, language='en'):
         return int(input_string)
 
     tokens = _tokenize(input_string)
-
+    lower_case_tokens = []
     for index, token in enumerate(tokens):
         compare_token = token.lower()
+        lower_case_tokens.append(compare_token)
         if compare_token in lang_data.all_numbers or compare_token.isspace() or len(compare_token) == 0:
             continue
         if compare_token in lang_data.skip_tokens and index != 0:
             continue
         return None
```
I would do:
```python
tokens = _tokenize(input_string)
normalized_tokens = _normalize_tokens(tokens)

for index, token in enumerate(normalized_tokens):
...
```

And extract the "`lower` logic" inside `_normalize`. We will probably need that function in the future to implement the Unicode normalization, so we can create it.

Tip: using `list(map(str.lower, tokens))`

Another way would be applying the `lower()`/normalization to the whole `input_string`, but I didn't try it and I don't know the consequences.

noviluni commented:
After fixing this I will help you with adding the missing supplementary data for Spanish :+1:

arnavkapoor commented:
Thanks a lot for all the inputs @noviluni , I will incorporate these changes. Any feedback with regards to the expected output for the test-cases in ```test_ambiguity_in_multipliers``` . I personally think they are giving the expected results. 
Also I believe you mentioned something about the import issue in the testing files ```from number_parser import parser``` , could you please retell that.

arnavkapoor commented on tests/test_language_en.py:
```diff
@@ -0,0 +1,145 @@
+            (100, 'hundred'),
+            (1000, 'thousand'),
+            (1000000, 'million'),
+            (1000000000, 'billion'),
```
Sure, I will update it. I am going with the basic English convention of grouping by 3. So, 1000 will be represented as 1_000 and similarly for larger numbers.

noviluni commented on tests/test_language_en.py:
```diff
@@ -0,0 +1,145 @@
+            ("million twenty three billion",
+                "1000023000000000"),
+
+            ("billion twenty three million",
+                "1000000023 1000000"),
```
Shouldn't this be `1023000000` (1_023_000_000)?

arnavkapoor replied:
Yes thanks, will need to fix it , I realize what is the mistake. 

Another thing this made me realize is that this part of the logic for deciding multiplier validity is dependent on how multipliers are spaced. eg) In English all multipliers are spaced by 10^3, 
``` thousand (10^3) ,  million (10^6),  billion(10^9).``` (However sometimes billion can mean (10^12) too.
 
However, I will change this and look to implement a language independent code. 
I will test this using ```en-IN``` locale as in India we have ```lakhs (1,00,000)``` and ```crore(1,00,00,000).``` Thus spaced by 10^2. 

For eg-
```en locale -> parse("one hundred thousand and fifty thousand") ==  "100050 1000"```
```en-IN locale -> parse("one lakh and fifty thousand") == 150000```

noviluni commented on tests/test_language_en.py:
```diff
@@ -0,0 +1,145 @@
+            ("two hundred hundred",
+                "200 100"),
+
+            ("million twenty three billion",
+                "1000023000000000"),
```
This case is weird to me... Not sure if we should add it here as the correct spelling is `one quadrillion twenty-three billion`...  Up to you.

arnavkapoor replied:
Agreed it's not the most natural sounding number, only similar sounding use case I have heard is in financial domains  ( five million billion dollar impetus to the economy etc ). But yeah doesn't really need to be here I guess.

noviluni commented:
@arnavkapoor 

> Any feedback with regards to the expected output for the test-cases in test_ambiguity_in_multipliers . I personally think they are giving the expected results.

I added two new comments. 

> Also I believe you mentioned something about the import issue in the testing files from number_parser import parser , could you please retell that.

Of course! Is not an issue in the tests, but an improvement of the UX (I know, using "UX" to refer to a library seems weird, you could call it "Developer Experience" if you prefer). 

What I was saying is that currently to use the `parse()` and the `parse_number()` functions we need to import them like this: `from number_parser import parser` and then `parser.parse()` or `from number_parser.parser import parse` and then `parse()`. 

As you can see this is really verbose. We are writing "parse" three times. I think we could improve it to work like this: `from number_parser import parse`. This can be achieved by:

1. Moving the `parse()` and `parse_number()` functions to the `__init__.py` file (this approach is used by `dateparser`, look at this: https://github.com/scrapinghub/dateparser/blob/master/dateparser/__init__.py)
2. Importing those functions inside the `__init__.py` file (this approach is used by Scrapy, for example, look at this: https://github.com/scrapy/scrapy/blob/master/scrapy/__init__.py).

Both are valid approaches. If you ask me I prefer the second approach because the code is better organized. So, in Scrapy, you can do: `from scrapy.http.request import Request` or `from scrapy import Request` and both work, but the code related to the requests is correctly organized in its own module.

Summarizing: You can add `from number_parser.parser import parse, parse_number` in the `__init__.py` file and we will be able to use `from number_parser import parse` (or `from number_parser import parse_number`, or even `import number_parser` and then: `number_parser.parse()`) directly. After that, you can change the import in the test files and use this in the examples in the docs.

Let me know if there is anything that it's not clear :)

arnavkapoor commented:
Hi, so I have been working on this PR and have faced a couple of issues and will list them and also the updates. I have incorporated all the mentioned changes and updated the supplementary files. At present all the cases for Russian and English work. Also the Spanish sentences mentioned [here](https://github.com/arnavkapoor/number-parser/pull/24#discussion_r454877698) have been added and are working. So the two major issues that have cropped up are as follows :- 

-  Adding large constants ```Gúgol``` , ```centillón``` in json supplementary file. From what i have looked into we can add large constants as ```1e600``` into the json however while reading them using ```write_complete_language_data``` script and merging into the final language file,  I am typecasting it to int there is rounding off issues. One solution would be to change all supplementary files from json to py but I think there should be an easier way without major changes. Ideas ?

- The only failing test (excluding cases from the above point) is ```cinco trillones setecientos sesenta y cuatro mil seiscientos siete billones quinientos mil millones treinta y uno```. Now this happens because the current logic on encountering a number greater than 1000 assumes that the number is completely formed and thus breaks the current group value. However, here when billiones comes we need to multiply it with the entirety of previous thing ```sete­cientos sesenta y cuatro mil seiscientos siete``` (764607). This basically happens due to [long and short scales](https://en.wikipedia.org/wiki/Long_and_short_scales), In Spanish because the next multiplier is ```10**6``` away your group value range increases from ```10**3``` in English to ```10**6```. This change I fell is best implemented by having scale information in the data file.

Now before implementing the 2nd change I think it's best to add the automated csv tests mentioned [here](https://github.com/arnavkapoor/number-parser/issues/15#issuecomment-659295377) so as to ensure that other things don't break and we have some confidence on the changes being made. They will also give us an idea of where we stand with the other languages.

arnavkapoor commented on number_parser/parser.py:
```diff
@@ -81,16 +115,18 @@ def _build_number(token_list, lang_data):
-            current_grp_value *= lang_data.big_powers_of_ten[token]
-            if not lang_data.big_powers_of_ten[token] == 100:
+            current_grp_value *= power_of_ten
+            if power_of_ten > 100:
```
This is basically group value being decided in case in English we see a power of 10 greater than 100. (i.e thousand , million , billion etc) we break. However for spanish this needs to ``` if power_of_ten > 100000``` as thousand can be used with a larger power later on.

noviluni commented on number_parser/parser.py:
```diff
@@ -47,29 +47,63 @@ def _check_validity(current_token, previous_token, lang_data):
+            for skip_token in used_skip_tokens:
+                value_list.append(skip_token)
+            used_skip_tokens = []
+            previous_power_of_10 = None
```
I'm not 100% confident, but it's a `continue` missing here? If that's not the case, you can remove the `used_skip_tokens = []` line as it's already reset below.

noviluni commented on number_parser/parser.py:
```diff
@@ -47,29 +47,63 @@ def _check_validity(current_token, previous_token, lang_data):
     return True
 
 
+def _check_large_multiplier(current_token, total_value, current_grp_value, lang_data):
```
Could you add a short docstring here? :)
